### PR TITLE
VideoPress:  add RN mobile video player block

### PR DESCRIPTION
**Pull request by jhnstn**
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5509

## Proposed changes:

* Add a native supported VideoPress video player block

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

TBD

### CODE REVIEW

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx [hunk 1]
```diff
@@ -39,6 +56,28 @@ export default function VideoPressEdit( { attributes, setAttributes } ): React.R
...
+		poster,
+	} );
+
+	// Get video preview status.
+	const { preview, isRequestingEmbedPreview } = usePreview( videoPressUrl );
+
+	// Pick video properties from preview.
+	const { html: previewHtml, scripts } = preview;
+
+	const html = previewHtml || cacheHtml;
```
Comment (fluiddot) [inline projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:79]: Regarding `cacheHtml` attribute, we'd need to replicate the same logic we have in the web version ([reference](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx#L160-L177)).

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	// Scale video player height based on the video ratio and layout width.
+	const scaleVideoPlayer = useCallback(
+		( { layout } ) => !! videoRatio && setVideoPlayerHeight( ( layout.width * videoRatio ) / 100 ),
+		[ videoRatio ]
+	);
+
+	return (
+		<View
+			style={ { height: videoPlayerHeight } }
+			onLayout={ ( { nativeEvent } ) => scaleVideoPlayer( nativeEvent ) }
```
Comment (fluiddot) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: I recall that the height was automatically set in the `Sandbox` component ([reference](https://github.com/WordPress/gutenberg/blob/b4e9e0aed4910eb7b6f57b67f866b37ea05e78ee/packages/components/src/sandbox/index.native.js#L264-L281)) communicating with the embed content, not sure if we could apply a similar approach for the VideoPress player 🤔 .

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * @returns {object}                     - React component.
+ */
+export default function VideoPressPlayer( {
+	html,
+	attributes,
+	scripts = [],
+	preview,
+	isRequestingEmbedPreview,
+} ) {
+	const [ videoPlayerHeight, setVideoPlayerHeight ] = useState( 250 );
```
Comment (fluiddot) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: [In the web version, a temporary height is set](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.js#L66-L73), maybe we could replicate a similar logic instead of having a constant, WDYT?

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		( { layout } ) => !! videoRatio && setVideoPlayerHeight( ( layout.width * videoRatio ) / 100 ),
+		[ videoRatio ]
+	);
+
+	return (
+		<View
+			style={ { height: videoPlayerHeight } }
+			onLayout={ ( { nativeEvent } ) => scaleVideoPlayer( nativeEvent ) }
+		>
+			{ embedPreviewRequested && <SandBox html={ html } scripts={ scripts } /> }
```
Comment (fluiddot) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: It's likely that we'd need to pass a custom JS script to communicate with the player, especially in relation to fetching the token for private videos. As an example, [here](https://github.com/Automattic/wp-calypso/blob/9b66da6ae2297cd78b19cad5982e66a516b2ca94/client/post-editor/media-modal/detail/detail-preview-videopress.jsx#L113-L146) is how Calypso handles the token request when asked from the player.

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx (hunk 1 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:79]: I'm not sure this is entirely true. 

I did see the effect that sets the cacheHtml attribute in the web version, which you referenced,  but I don't see where the attribute is actually being used in the video player component.  For now I would suggest we keep the current logic scoped to what is needed to render the video player.

That said I find the comments above the effect hook puzzling and makes wonder if the `cacheHtml` attribute is actually needed. The comment mentions setting [`html` and `thumbnail` attributes](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx#L162), neither of which are actually set in the effect hook. 
I'll do some commit spelunking to find the reasoning for the setting the `cacheHtml` attribute, perhaps it is needed for the some of the other components like `<VideoPressUploader>` or `<PlaceholderWrapper>`

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 2 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: I _believe_ I was running into issues with the aspect ratio of the videos, not just the height. I'll play around with this to see if I can get the Sandbox to handle it correctly

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 3 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: Are you referencing the magic number `250` ?  That's just a replacement for the temporary styling height that was added [here](https://github.com/Automattic/jetpack/pull/29165/files#diff-a48bfa252db3a7399b5a161b6b36a5b88c3db3d37780daece46693f418238647L12). I would expect that to be modified later once we get into refining the component.  That said the web version [also uses a constant](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.js#L189), just not in the `useState` call. ( Just noticed while writing this that the web version now uses [a constant as well](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.tsx#L71) )

Or are you talking the web version use of  `state.videoPlayerTemporaryHeight` . It's really just naming.  I've simplified the logic involved in calling `setVideoPlayerHeight` vs the web version `setVideoPlayerTemporaryHeight` but the intent is the same. I expect the calling logic to become more complex in future changes so I opted to keep it simple for now. 
I found the use of _temporary_ misleading in the web version. I couldn't find a condition where the "temporary" height is not used, it is ignored once the embed is rendered but it's not replaced or removed. Perhaps a better name, in both cases, is `playerWrapperHeight` since that is where how the state property is [used](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.tsx#L201).

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 4 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: Agreed, however that feels like a conversation for future changes.
The scope here is to get a native version of the player working for public videos.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 2 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: Ok I figured out what's going on here. I was following the web version a bit too closely I think. There is a "loading" view that needs some height.  I think the web version also needs to handle responsive UI changes so I'm assuming the height needs to be handled explicitly.

I can handle the "is loading" state better which simplifies a lot of things. 
Thanks for pointing that out!

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 4 above)
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: Good point. I agree that if we limit the scope to public videos this won't be necessary 👍. Let's keep this in mind in the next iteration but not as a blocker of the PR.

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx (hunk 1 above)
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:79]: > I did see the effect that sets the cacheHtml attribute in the web version, which you referenced, but I don't see where the attribute is actually being used in the video player component. For now I would suggest we keep the current logic scoped to what is needed to render the video player.

I think the purpose of the `cacheHtml` attribute is to be used upon the component being mounted to display the video player while the preview is being fetched, as it's mentioned in this comment:
https://github.com/Automattic/jetpack/blob/46d3698a631903c49531e264dcb18f35816d7d40/projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx#L184-L186

I understand that this is an enhancement to avoid displaying a loading indicator, which we might consider implementing in the native version. In any case, I tested this functionality in the web version and it's true that is hard to tell the difference with and without having the cache, especially as the HTML content is an iframe that still needs to load the video player content 🤔.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 3 above)
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: > Are you referencing the magic number `250` ? That's just a replacement for the temporary styling height that was added [here](https://github.com/Automattic/jetpack/pull/29165/files#diff-a48bfa252db3a7399b5a161b6b36a5b88c3db3d37780daece46693f418238647L12). I would expect that to be modified later once we get into refining the component. That said the web version [also uses a constant](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.js#L189), just not in the `useState` call. ( Just noticed while writing this that the web version now uses [a constant as well](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.tsx#L71) )

Yeah, I was referencing the magic number. Oh, true, in the web version we also have a constant. In that case, omit my comment then, thanks for elaborating!

> Or are you talking the web version use of `state.videoPlayerTemporaryHeight` . It's really just naming. I've simplified the logic involved in calling `setVideoPlayerHeight` vs the web version `setVideoPlayerTemporaryHeight` but the intent is the same. I expect the calling logic to become more complex in future changes so I opted to keep it simple for now.
> I found the use of _temporary_ misleading in the web version. I couldn't find a condition where the "temporary" height is not used, it is ignored once the embed is rendered but it's not replaced or removed. Perhaps a better name, in both cases, is `playerWrapperHeight` since that is where how the state property is [used](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.tsx#L201).

Based on the following comment, I understand that the temporary height is just an initial calculation we make based on the block's width and the video's aspect ratio. Once the player is loaded, the height is replaced with the final value, so it seems the temporariness is based on the fact that is only used during the loading state.

https://github.com/Automattic/jetpack/blob/8f7114bde33077e8b45940ac579126c4d77a2ac8/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.tsx#L60-L66

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx (hunk 1 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:79]: > I think the purpose of the cacheHtml attribute is to be used upon the component being mounted to display the video player while the preview is being fetched

Seems like the `const html  = previewHtml || cacheHtml` assignment covers the mounting period in the block lifecycle. 

This enhancement might be for when the preview is fetched _after_ the component is mounted. The hook logic suggests that the `previewHtml` would have to exist then not exist before the `setAttribute` call would have any relevance.  

I'm guessing that happens when the video source is updated. If so, we should probably handle that when we get to settings or uploading updates.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js (hunk 3 above)
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: It's true that there is an initial value for the height that is derived from the aspect ratio. 
The issue I see is the naming. From what I observed the `videoPlayerTemporaryHeight` state variable is always used to set the style height property. 

In fact once the video is available, `videoPlayerTemporaryHeight` is set to `auto` [here](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.js#L103) 

In other words, this state variable has an initial value to set a placeholder height, and then is set to `auto` once the actual height is available. To me that variable is not temporary. If I see something named temporary with a comment that it's eventually ignored,  I would expect to see it used like 

```
const value = actualValue || temporaryValue
```

instead the code here is more like 
```
let temporaryValue = 1;
eventually( (actualValue) => setTemporaryValue(actualValue) )
const value = temporaryValue
```

Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.native.js]: > In fact once the video is available, `videoPlayerTemporaryHeight` is set to `auto` [here](https://github.com/Automattic/jetpack/blob/36fde37a4444d1d1debcef9d5b3bae8951aa1dc7/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-player/index.js#L103)
> 
> In other words, this state variable has an initial value to set a placeholder height, and then is set to `auto` once the actual height is available. To me that variable is not temporary. If I see something named temporary with a comment that it's eventually ignored, I would expect to see it used like

Yep, good point. I agree the `videoPlayerTemporaryHeight ` is a bit confusing, especially when the `auto` value is set because is not consistent with the temporariness led by the variable name. I understand that the purpose of `auto` is basically disabling the temporary height and letting the video player control the height, maybe the value should be renamed to `disabled` or something similar 🤔. For the native version, if we have to provide a customized version of the component, we can discuss further how to name and structure this part if needed.
